crates/typst-library/src/foundations/int.rs

```diff
@@ -42,26 +43,73 @@ type i64;
 #[scope]
 impl i64 {
     /// Converts a value to an integer. Raises an error if there is an attempt
-    /// to produce an integer larger than the maximum 64-bit signed integer
-    /// or smaller than the minimum 64-bit signed integer.
+    /// to parse an invalid string or produce an integer that doesn't fit
+    /// into a 64-bit signed integer.
     ///
     /// - Booleans are converted to `0` or `1`.
     /// - Floats and decimals are rounded to the next 64-bit integer towards zero.
-    /// - Strings are parsed in base 10.
+    /// - Strings are parsed in base 10 by default.
     ///
     /// ```example
     /// #int(false) \
     /// #int(true) \
     /// #int(2.7) \
     /// #int(decimal("3.8")) \
-    /// #(int("27") + int("4"))
+    /// #(int("27") + int("4")) \
+    /// #(int("beef", base: 16))
     /// ```
     #[func(constructor)]
     pub fn construct(
         /// The value that should be converted to an integer.
-        value: ToInt,
-    ) -> i64 {
-        value.0
+        value: Spanned<ToInt>,
+        /// The base (radix) for parsing strings, between 2 and 36.
+        #[named]
+        #[default(Spanned::new(Base::Default, Span::detached()))]
+        base: Spanned<Base>,
+    ) -> SourceResult<i64> {
+        Ok(match value.v {
+            ToInt::Int(n) => {
+                if matches!(base.v, Base::User(_)) {
+                    bail!(base.span, "base is only supported for strings");
+                }
+
+                n
+            }
+            ToInt::Str(s) => {
+                let b = base.v.value();
+                if !(2..=36).contains(&b) {
+                    bail!(base.span, "base must be between 2 and 36");
+                }
+
+                // Replace minus sign at the start with `-`
+                let s: &str = if s.as_str().starts_with(repr::MINUS_SIGN) {
```

Comment:
this is a bit inefficient. could you we do the sign handling separately to avoid the allocation? (I presume we might run into issues with max/min integers like we did before?)